File: backend/src/AnimeShop/API/Admin/Units.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Admin.Units where

import AnimeShop.Prelude

type UnitsAPI =
  "units" :>
  (
    Get '[JSON] [Entity Unit] :<|>
    ReqBody '[JSON] Unit :> Post '[JSON] (Entity Unit) :<|>
    ReqBody '[JSON] UnitId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity Unit) :> Patch '[JSON] ()
  )

unitsServer :: ServerT UnitsAPI App
unitsServer = getUnits :<|> postUnit :<|> deleteUnit :<|> patchUnit

getUnits :: App [Entity Unit]
getUnits = runDb $ selectList [] []

postUnit :: Unit -> App (Entity Unit)
postUnit = runDb . insertEntity

deleteUnit :: UnitId -> App ()
deleteUnit = runDb . delete

patchUnit :: Entity Unit -> App ()
patchUnit entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/units.js
import { ADD_UNIT, GET_UNITS, DELETE_UNIT } from '../actions/types';

export function units(state = [], action) {
  switch (action.type) {
    case ADD_UNIT:
      return [ ...state, action.unit];
    case GET_UNITS:
      return action.units;
    case DELETE_UNIT:
      return state.filter(unit => unit.id !== action.id);
    default:
      return state;
  }
}
